### ISSUE TITLE
Separate SVG parser into standalone library

### ISSUE THREAD

**Issue by notjosh** [mogest/prawn-svg#65]
Hi! I stumbled on the project, and it seems really promising and cool. Great work so far!

I don't need PDF writing, nor do I care much for Prawn. However, I'd really love to have a standalone SVG parser in Ruby. (I want to write out to other format/s.)

I think it's feasible to break `prawn-svg` into two components - one to read the SVG into its core elements (which can exist standalone!), and one to convert the nodes into Prawn calls.

I don't want to tinker too much, as it seems you're active in dev. But looking at master, it appears to be in pretty good shape to do exactly what I'm proposing - and maybe that's the direction you're already heading :)

Thoughts?

**Comment by mogest** [comment]
Hey @notjosh.  Yeah, I structured it in two parts so that the element renderer didn't know anything about prawn.  It builds a call tree in memory, and then executes the call tree in the prawn interface.  But the calls are prawn-specific at the moment; I didn't think there was much benefit to invent an entirely new intermediary language.

A lot of the processing it does is specifically to get it into a state that is renderable by PDF, too.  For example PDF doesn't have the concept of an arc, so I have to approximate one using bezier curves.

Would be very interested to hear how your thoughts about how we could split it into two.

**Comment by notjosh** [comment]
Yeah! Even though it's not prawn-specific, it's still very married into PDF and flattened. The result is a flat call tree, with no hierarchy. I need/want that hierarchy! :)

And then, different readers should be able to parse the attributes however the destination format needs it. i.e. if the node tree kept the `path` element attributes, then the PDF reader could parse into bezier curves, but other readers could parse it into whatever native format is best for its destination.

So I think there are a few steps this can be broken into. You've done the hard work of the parsing, creating node entities, etc. So I think the first step is to simply retain the document/node tree in parallel with the calls, complete with attributes. (So, in `Prawn::SVG::Elements::Base` instead of just `:parent_calls, :base_calls`, we'd add `:children`, which `process_child_elements` would write into.)

I think that'd be enough for me to be able to parse the tree in Ruby, and then write out my own non-PDF calls.

Then, it should be easy enough to pull the PDF calls out into a separate reader that reads a parsed  `Document` externally. You have the code for it, it's just a matter of splitting the classes :)

**Comment by mogest** [comment]
Aha!  I understand what you're after.  Coincidental that I only set up this new element structure a week ago, huh.

Leave it with me for a bit, I'll see what I can do.  Got any particular time constraints on your side?

**Comment by notjosh** [comment]
Cool!

Nothing _super_ urgent, but we're shipping something in the next couple of weeks that could benefit. It's not a blocker though, so whatever works for you :) Happy to offer up time, too, but I feel I'd be treading on your toes.

**Comment by mogest** [comment]
Let me have a bash at it first.  Probably won't be release-ready in two weeks, but we'll see how it goes.

**Comment by notjosh** [comment]
Did you make any progress with this? I've hacked on a branch and totally destroyed the Prawn integration, but the SVG tree is nice and independent. I'm writing it out to generate paths for drawing on iOS now - it's really cool! :)

**Comment by mogest** [comment]
Sorry I didn't.  I sat down one afternoon and started to map out what needed to be done, and then just stopped when I realised what I'd need to do to maintain backwards compatibility as well as do a good job of it!  Possible but time consuming...

That's great that you hacked it up to work though!  I'll keep it in mind in case I have an urge to do a really huge refactor one day (and will keep this open too.)

**Comment by mogest** [comment]
I'm doing some github housecleaning and came across this one.  prawn-svg has moved more in this direction in the last three years but I don't think it'll ever reach there; I guess it's not the purpose I intend for the library, and with the current trajectory of how ruby is being used by the industry I don't think there'd be a huge demand for it.

So sadly I will close this for now.  Thanks for the suggestion tho, hope your hack is still working for you!
